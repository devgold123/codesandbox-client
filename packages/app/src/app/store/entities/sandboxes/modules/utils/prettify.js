import { DEFAULT_PRETTIER_CONFIG } from 'app/store/preferences/reducer';

// function getParser(mode) {
//   if (mode === 'jsx') return 'babylon';
//   if (mode === 'json') return 'json';
//   if (mode === 'css') return 'postcss';
//   if (mode === 'html') return 'parse5';
//   if (mode === 'typescript') return 'typescript';
//   if (mode === 'graphql') return 'graphql';

//   return 'babylon';
// }

function getMode(title: string) {
  if (/\.jsx?$/.test(title)) {
    return 'babylon';
  }

  if (/\.tsx?$/.test(title)) {
    return 'typescript';
  }

  if (/\.json$/.test(title)) {
    return 'json';
  }

  if (/\.css$/.test(title)) {
    return 'postcss';
  }

  if (/\.scss$/.test(title)) {
    return 'postcss';
  }

  if (/\.less$/.test(title)) {
    return 'postcss';
  }

  if (/\.gql$/.test(title)) {
    return 'graphql';
  }

  return null;
}

export function canPrettify(title) {
  return !!getMode(title);
}

let worker = null;

export default function prettify(
  title,
  code,
  prettierConfig = DEFAULT_PRETTIER_CONFIG
) {
  const mode = getMode(title);

  worker = worker || new Worker('/static/js/prettier/worker.js');

  return new Promise(resolve => {
    if (!mode) {
      resolve(code);
      return;
    }

    worker.postMessage({
      text: code,
      options: {
        ...DEFAULT_PRETTIER_CONFIG,
        ...prettierConfig,
        parser: mode,
      },
    });

    let timeout = setTimeout(() => {
      // If worker doesn't respond in time
      resolve(code);
    }, 5000);

    worker.onmessage = e => {
      const { formatted, text, error } = e.data;
      if (timeout) {
        clearTimeout(timeout);
        timeout = null;
        if (text === code) {
          if (error) {
            console.error(error);
            resolve(text);
          }
          if (formatted) {
            resolve(formatted);
          } else {
            resolve(text);
          }
        }
      }
    };
  });
}
