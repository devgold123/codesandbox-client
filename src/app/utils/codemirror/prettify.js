export default (async function prettify(code, eslintEnabled) {
  const prettier = await System.import('prettier');
  const prettifiedCode = prettier.format(code, {
    singleQuote: true,
    trailingComma: 'all',
  });

  if (eslintEnabled) {
    const { fix } = await System.import('app/utils/codemirror/eslint-lint');
    try {
      const lintedCode = fix(prettifiedCode).output;

      // If linter can't parse, it will return an empty string. Unwanted, so fall
      // back to prettified code
      if (lintedCode) return lintedCode;
    } catch (e) {
      console.error(e);
    }
  }

  return prettifiedCode;
});
