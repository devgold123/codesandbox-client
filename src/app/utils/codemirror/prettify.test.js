import prettify from './prettify';

const testCode = async (code: string) => {
  const prettified = await prettify(code);
  expect(prettified).toMatchSnapshot();
};

describe('prettify', () => {
  it('should prettify code', async () => {
    await testCode(`const a='32'`);
  });

  it('should prettify static class props', async () => {
    await testCode(
      `
      class Ives extends React.Component {
        property = {

        }
      }
    `,
    );
  });

  it('should prettify syntax errors', async () => {
    await testCode(`apofdawoifefjaweoifj`);
  });
});
