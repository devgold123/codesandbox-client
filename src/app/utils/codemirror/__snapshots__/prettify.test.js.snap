// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prettify should prettify code 1`] = `
"const a = '32';
"
`;

exports[`prettify should prettify static class props 1`] = `
"class Ives extends React.Component {
  property = {};
}
"
`;

exports[`prettify should prettify syntax errors 1`] = `
"apofdawoifefjaweoifj;
"
`;
